### ISSUE TITLE
Cheaters can quickly put yourself in front of the blocks

### ISSUE THREAD

**Issue by Skilzar** [NoCheatPlus/Issues#369]
#### Short description of the issue.

Hello! :)
Cheaters can quickly put yourself in front of the blocks
Video: https://www.youtube.com/watch?v=wH138HC1wRg

#### Complete output of the _ncp version_ command (please prefer text here).

---- Version information ----
#### Server ####
git-Spigot-21fe707-e1ebe52 ~MC: 1.8.8~
  detected: 1.8.8
#### NoCheatPlus ####
Plugin: 3.15.0-SNAPSHOT-sMD5NET-b1054
MCAccess: 1.8.4-1.8.8 / Spigot-CB-1.8_R3
Features:
  blocks: BlocksMC1_4 | BlocksMC1_5 | BlocksMC1_6_1 | BlocksMC1_7_2 | BlocksMC1_8
  checks: FastConsume | Gutenberg | AttackFrequency | FlyingFrequency | KeepAliveFrequency
  defaults: pvpKnockBackVelocity
  packet-listeners: UseEntityAdapter | MovingFlying | OutgoingPosition | KeepAliveAdapter | SoundDistance | CatchAllAdapter
Hooks: AllViolations~NCP~ 1.0
#### Related Plugins ####
ProtocolLib v4.1.0

#### How to reproduce the issue (is it random / always happens / side conditions)?

Cheat

#### Have you made any (noteworthy) changes to the default configuration of NCP?

There is, but only on the blocked commands

#### Do you run any other special plugins that affect game mechanics next to NC+ (adding skills, machines, adding/changing blocks, other anti cheating)?

ViaVersion

**Comment by Janmm14** [comment]
Also known as ScaffoldWalk

**Comment by jonesdj1** [comment]
you can also do this legit

**Comment by Janmm14** [comment]
Hack clients can do this much faster and from impossible angles.

**Comment by jonesdj1** [comment]
@Janmm14 there is nothing impossible about the angles

**Comment by Janmm14** [comment]
@jonesdj1 I mean as far as I know NoCheatPlus allowes placing blocks on another block without looking to the correct side of the block.

**Comment by MarkElf** [comment]
The output of /ncp debug player (player name) would probably help determine whether the player is teleporting a small portion, likely via packets similar to blink, or moving quicker through a different method. 

https://github.com/NoCheatPlus/Docs/wiki/Debugging

**Comment by asofold** [comment]
A debug log with ProtocolLib and debugging activated by config hehe :) - that would be best (global debugging isn't nice with production servers / with more than one player).

In general this issue doesn't look like all too severe, i can imagine you'll be faster with using a hack. Concerning the block face and the looking direction, NCP has various checks on that account, i'm not sure why none would trigger - one explanation could be that the client actually does the necessary moves, but you don't see it, because they're sent fast or with odd timing/sequence. Possibly it's also a weakness in NCP, though the targeted block is not the eye block, so i'm not so sure what's this about (blockinteract.visible, and direction fpr both blockinteract and blockplace). Looks like a block face check is not included in the direction check(s), so it would be the job of the visible check, and the block that is in the way is the targeted block, just with the wrong block face - that could be the reason.

**Comment by Janmm14** [comment]
@asofold
Yeah the clients do send different yaw/pitch, but to my knowledge it is not required to be on the correct side of the block, just looking at it allowes placing in any direction to the block.
Some people use modes where the blocks are built 2-3 blocks before your feet also.
Additionally enforcing the limit can cause block glitches for players who use things like trying to "noshift" such a way, so one might want to check wherher the next movement after the block place would allow sich a block place (I think it could also be the client which could send block places before movements)

**Comment by asofold** [comment]
The blockinteract.visible ray-tracing still uses an overly forgiving implementation. It's been planned to be replaced by a passable-like thing a while ago, however such needs quite some testing to stabilize, so it's been queued since.

Perhaps one could  make the pre-checks detect wrong block faces better, including ensuring the pre-checks run at all for this case.

**Comment by Skilzar** [comment]
[nocheatplus.log.zip](https://github.com/NoCheatPlus/Issues/files/908203/nocheatplus.log.zip)
Debug how quickly put the blocks under him

**Comment by asofold** [comment]
For the "building your own ground" thing, it looks like they do it at walk speed.
* Walk speed ensures a safe maximum thing there.
* (Sprint could cause sprint-backwards check, or if done too fast with turning around, it could trigger morepackets.)
* (Not checked for 'speed up by sending more packets', yet.)

In theory this might even be possible, of course you wouldn't see it with 100% accuracy. So this is a little bit problematic to block. Certainly an optional heuristic detection could be made (alert/block would be configurable, not sure what to  make the default).

Currently one could also let it feed the improbable check, if we detect such a mixture of taken distance vs. walking on just placed blocks, so it could penalize extreme speeds.
